api/v1beta2/ratelimitpolicy_types.go

```diff
@@ -106,7 +106,7 @@ type Limit struct {
 }
 
 func (l Limit) CountersAsStringList() []string {
-	var ret []string
+	ret := make([]string, 0)
```

Comment:
These 2 ways to declare the slice are NOT equivalent. Even though I think we don't usually rely on null check for slices, but rather on `len(s) == 0`, effectively this change alters the behaviour of the function and should be avoided.

If the goal is to improve memory allocation and performance, then I would do:

```go
size := len(l.Counters)
if (size == 0) {
	return nil
}
ret := make([]string, len)
// we cannot use copy(ret, l.Counters) due to different types
for idx := range l.Counters {
	ret[idx] = string(l.Counters[idx])
}
return ret
```

Or, in this particular case:

```go
size := len(l.Counters)
if (size == 0) {
	return nil
}
return common.Map(l.Counters, func(counter ContextSelector) string { return string(counter) })
```

Reply:
Thanks ! I've pushed this change 👍

Reply:
The pattern is also applied in https://github.com/Kuadrant/kuadrant-operator/pull/230/commits/4bf4198540c75fc37d44ebf8d1f080ec1a6762d0#diff-3565ea71e4b2b9f8f80ea1c57659cc4bf19888115b8ccca64958f797eb39bcd4R28

Reply:
Should the slice size check be applied also  ?

```go
	if len(rlpRefs) == 0 {
		return nil
	}
```
In this instance the logic from `r.reconcileLimitador(ctx, rlp, rlpRefsWithoutRLP)` will be skipped ? 🤔 
- https://github.com/KevFan/kuadrant-operator/blob/rlp-v2-rebase/controllers/ratelimitpolicy_limits.go#L35

I guess the `Filter` function should be used instead ? Though it looks to set the capacity also to `make([]T, 0)`
- https://github.com/KevFan/kuadrant-operator/blob/rlp-v2-rebase/pkg/common/common.go#L159-L168

Something like:
```go
	rlpRefsWithoutRLP := common.Filter(rlpRefs, func(rlpRef client.ObjectKey) bool {
		return rlpRef.Name != rlp.Name && rlpRef.Namespace != rlp.Namespace
	})
```

Reply:
> Should the slice size check be applied also ?

I would say not. Let it pass `nil` into `reconcileLimitador` so we do not change the behaviour.

<br/>

> I guess the `Filter` function should be used instead ? Though it looks to set the capacity also to `make([]T, 0)`

Yes, but watch out for the logic though. The `Filter` function receives a predicate function that appends the item to the resulting slice when true. In this case, we want to append all `rlpRef`s but the one whose name and namespace match `rlp`'s name and namespace. ∴

```go
func (r *RateLimitPolicyReconciler) deleteLimits(ctx context.Context, rlp *kuadrantv1beta2.RateLimitPolicy) error {
	rlpRefs, err := r.GetAllGatewayPolicyRefs(ctx, &common.KuadrantRateLimitPolicyRefsConfig{})
	if err != nil {
		return err
	}
	rlpRefsWithoutRLP := common.Filter(rlpRefs, func(rlpRef client.ObjectKey) bool {
		return rlpRef.Name != rlp.Name || rlpRef.Namespace != rlp.Namespace
	})
	return r.reconcileLimitador(ctx, rlp, rlpRefsWithoutRLP)
}
```

Reply:
I see, thanks ! Updated 👍